### ISSUE TITLE
`lint` passes over a stale `file()` ref that `compile` already wrote to disk

### ISSUE THREAD

**Issue by vlad-ryzhkov** [zernie/agent-lint#173]
Tested on `vigiles@21.0.1` (npm) = `681edf7`.

`README.md:166` says: "**`lint` is the CI gate** that fails the build on the same deterministic
checks — broken refs, bad tool contracts, dead hooks, skill collisions". `command: lint` is the
default for the GitHub Action. For spec mode that promise does not hold.

## Repro

Empty directory, `vigiles@21.0.1`:

```ts
// CLAUDE.md.spec.ts
import { instructionFile, file, cmd } from "vigiles/spec";
export default instructionFile({
  sections: {
    Overview: `Gate reproduction.`,
    Routing: [file("docs/never-existed.md"), cmd("npm run never-existed")],
  },
  rules: {},
});
```

```
$ npx vigiles compile CLAUDE.md.spec.ts
✗ CLAUDE.md.spec.ts — 2 error(s)
  [stale-file] File not found: "docs/never-existed.md"
  [stale-command] Script "never-existed" not found in package.json
Compilation complete with errors.          # exit 1

$ cat CLAUDE.md
<!-- vigiles:sha256:b137ea1d13bccd39 compiled from CLAUDE.md.spec.ts -->
...
`docs/never-existed.md``npm run never-existed`

$ npx vigiles lint .
  ✓ CLAUDE.md — hash valid (from CLAUDE.md.spec.ts)
  All compiled files intact.               # exit 0
```

## Why it happens

Two things stack:

1. `compile` prints the errors, returns exit 1, and **still writes the artifact** — with a valid
   sha header. A hash-valid file holding known-dead refs is a false-confidence artifact by
   definition.
2. `lint` trusts the hash. `checkIntegrity` (`src/core/integrity.ts:120`) compares the file body
   against its own header and never re-derives the refs from the spec.

Who hits this: you compile locally, see the error, get distracted, commit. CI runs `vigiles lint`
— the command the docs call the gate for broken refs — and it is green.

## Patch — one deleted line

`src/cli.ts:498-503`, the error branch of `compileClaudeToFile`:

```ts
  if (errors.length > 0) {
    console.log(`\n✗ ${specPath} — ${String(errors.length)} error(s)`);
    printErrors(specPath, errors);
    writeFileSync(resolve(basePath, primaryOutput), markdown);   // <- delete line 501
    return false;
  }
```

Deleting line 501 means a spec with stale refs produces no artifact. The error stays visible, exit 1
stays, and a hash-valid file with dead refs is no longer emitted.

Plus one documentation line: `README.md:166` either narrows "broken refs" to what `lint` really
checks, or states that spec mode needs `compile` in CI, not `lint`. Right now the promise is wider
than the implementation.

## What the patch does not close

An already-committed poisoned `CLAUDE.md` stays green under `lint`, because `checkIntegrity` hashes
the file on its own. Closing that fully means re-verifying spec refs in `lint` when a `.spec.ts`
sits next to the output. That is more than one line and your call. Deleting line 501 stops new
artifacts of this kind and costs nothing.

**Comment by zernie** [comment]
Auto-closed by #186 (`94af23a5`) — leaving the reasoning here, since the merge said nothing.

**What shipped.** Your one-line patch, plus the doc half you asked for. `compile` no longer writes an artifact when references are dead:

```
✗ CLAUDE.md.spec.ts — 2 error(s)
  [stale-file] File not found: "docs/never-existed.md"
  [stale-command] Script "never-existed" not found in package.json
  → CLAUDE.md was NOT written; the previous version is left in place.
```

The last good artifact surviving is strictly better than being replaced by a broken one — the error stays on screen, exit 1 stays, and no valid hash is minted over known-dead refs.

**Measured.** Three e2e cases through the real built CLI (`src/cli-compile-gate.test.ts`), one of which is your exact repro; a mutation restoring the write fails all three. The middle case is the one your patch made me think about: a *previously good* artifact must survive a failed recompile untouched, and `lint` must still be green on it — refusing to write must not mean destroying what is there.

**The doc line.** `README.md:166` is narrowed rather than left overstated: with a spec, `compile` is what re-derives refs and `lint` verifies the compiled file is intact. Your framing — "the promise is wider than the implementation" — is the accurate one.

**What is still open, and you named it yourself.** An already-committed poisoned `CLAUDE.md` stays green under `lint`, because `checkIntegrity` hashes the file against its own header and never re-derives refs from the spec. Deleting the write stops *new* artifacts of this kind; it does not clean up an existing one. That closure is a real change (re-verifying spec refs during `lint` when a `.spec.ts` sits beside the output) and did not ship here. Worth its own issue if you want it tracked — I would rather leave that decision to the maintainer than file scope on their behalf.

Thank you for the repro. An empty directory, a five-line spec and three commands is the cheapest possible bug report, and it took about a minute to confirm.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by zernie** [comment]
Follow-up: **the residual you named is now closed too**, in #187 (`85c54e7e`).

When this was first fixed I only stopped `compile` from writing new poisoned artifacts, and quoted your own caveat back at you — that an already-committed `CLAUDE.md` stays green because `checkIntegrity` hashes the file on its own. I said it was more than one line and left it. It was the half that mattered, so here it is.

**The hole, reproduced:**

```
$ vigiles compile CLAUDE.md.spec.ts
✗ [stale-file] File not found: "docs/guide.md"

$ vigiles lint .
✓ CLAUDE.md — hash valid (from CLAUDE.md.spec.ts)
  All compiled files intact.                        # exit 0
```

Two different claims were being read as one. A hash answers *is this still what the spec compiled to*; it says nothing about whether the things the file **names** still exist. An artifact committed while its refs were live stayed green forever after the target was deleted.

**The new `spec-refs` rule** re-derives a compiled file's references from its spec and validates them exactly as `compile` does. Default `error`, matching `compile` — a dead path is decidable from the filesystem, not a proxy, so it belongs in the hard tier. Only specs whose output already exists are loaded, so a repo with no specs pays nothing.

🔴 **It immediately found four real dead references in this repository.** `examples/CLAUDE.md.spec.ts` still pointed at `src/spec.ts`, `src/compile.ts`, `src/linters.ts` and `src/generate-types.ts` — all four moved to `src/core/` in the hexagonal refactor and the example's spec was never updated. Exactly the class you filed, sitting in the home repo the whole time, green.

⚠️ **And an existing test asserted the opposite.** `a failed compile leaves the LAST GOOD artifact alone` also claimed `lint` must be GREEN on that survivor — I wrote that assertion in the first fix. It *was* the hole, written down as an expectation. It now asserts exit 2 plus the named reference; the part that matters (refusing to write must not destroy what is there) is still pinned by comparing bytes.

`README.md`'s "lint is the CI gate … broken refs" line is finally true for spec mode rather than narrowed around the gap.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by zernie** [comment]
**Correction to my previous comment.** I told you the residual was closed. It was closed for `CLAUDE.md` and open everywhere else, and I did not check before saying so.

`compileClaudeToFile` stopped writing on error. The identical three lines — write, **then** check errors — sat unchanged in the skill, subagent, railway and generator compilers, and all four went on stamping unconditionally. Worse than the original case: the `spec-refs` rule I pointed at as the backstop skips non-`claude` specs, so for those four there was neither a write-gate nor a lint-side catch.

Your original repro, moved one file sideways:

```
$ vigiles compile skills/demo/SKILL.md.spec.ts
✗ [stale-file] File not found: "docs/never-existed.md"
skills/demo/SKILL.md — 256 bytes, stamped

$ vigiles lint .
✓ skills/demo/SKILL.md — hash valid (from skills/demo/SKILL.md.spec.ts)
0 finding(s) — exit 0
```

That is the bug you filed, still live, on the day I said it was gone.

**What shipped now** (#188, `7693c71`): not four more guards. `addHash` mints a branded `StampedMarkdown`, `seal()` hands one out only when `errors` is empty, and `writeArtifact` accepts nothing else — so a failing compile has no stamp to give the writer, in any of the five, and the sixth compiler will be written the same way. The unstamped `markdown` string stays available because `adoptDiff` legitimately wants a draft to diff against even when a ref is dead.

**Measured.** An e2e case per compiler through the built CLI; a mutation restoring write-before-check in the skill compiler fails on `a SKILL spec with a dead ref writes nothing either`. ⚠️ That mutation came back **green** on its first run — it had landed on the first of four identical blocks instead of the one under test. Worth saying out loud, since it is the same failure mode as the comment I am correcting: a check that looks like it ran.

**What I am not claiming.** The write path is closed across all five; `spec-refs` still only re-derives refs for `claude` specs, so an already-committed poisoned `SKILL.md` from an older version stays green under `lint`. That is the same residual as before, now scoped honestly, and it is filed separately rather than asserted fixed here.

---
_Generated by [Claude Code](https://claude.ai/code)_
